### ISSUE TITLE
Generate needs to consider trainer model context window length

### ISSUE THREAD

**Issue by HunterGerlach** [instructlab/instructlab#1024]
When running `ilab generate` on a qna.yaml where questions/answers are long, the command can fail due to the trainer model's context window limitations.

`Generating dataset failed with the following error: There was a problem connecting to the server Error code: 400 - {'error': {'message': "This model's maximum context length is 4096 tokens. However, you requested 9360 tokens (6288 in the messages, 3072 in the completion). Please reduce the length of the messages or completion.", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}`

To reproduce this, use the base instructions for creating a custom model (instructlab/README.md), create a qna.yaml where the answers are really verbose, which should encourage the model to replicate the verbose pattern, and run `ilab generate`. After a few synthetic pairs you should see a failure.

If normal q/a pairs don't work, try replicating the same word over and over for multiple lines for each answer. This seems to encourage the model to be extremely verbose (and the content doesn't really matter for this test).

Also, don't forget to include the prompt.txt in the context size as I believe it also gets sent along with the qna content.

**Comment by bbrowning** [comment]
I frequently hit this as well (with knowledge generation in my case), where my `ilab generate` run eventually errors out with a similar context length exceeded message. Perhaps we need to check our generated prompts before sending to the model and skip anything we generate that's too long? That would be better than exiting the entire generation run with this error.

**Comment by gshipley** [comment]
I am also hitting this with knowledge and skills.

**Comment by aakankshaduggal** [comment]
Yes @gshipley, ideally we would want to keep the tokens less than 4096.

Can you point me to the taxonomy yaml you are trying to run?

**Comment by gshipley** [comment]
I am using the qna.yaml file here:
https://github.com/gshipley/softball-instruct/blob/main/qna.yaml

I also just saw in config.yaml and ilab serve --help the following:

--max-ctx-size INTEGER  The context size is the maximum number of tokens
                          considered by the model, for both the prompt and
                          response. Defaults to 4096.

I assume I can use this as a workaround. Set it 20000 or something?

**Comment by aakankshaduggal** [comment]
I wouldn't recommend increasing the tokens. Rather chunking the document is a better approach. Ideally the ilab should be chunking the document into lesser number of tokens to avoid this issue. I see https://github.com/instructlab/instructlab/pull/969/ tried to address this issue.

**Comment by aakankshaduggal** [comment]
We are working on getting https://github.com/instructlab/instructlab/pull/969/ merged and @gshipley you can try after the new updates are made, that should resolve the issue you are facing.

**Comment by bbrowning** [comment]
969 won't fix this issue unless it's specifically because there are really long lines in the knowledge document. I see some that are a bit long there, but nothing that I'd think would be long enough to trigger this. The error @gshipley got is over the context window by about 3000 tokens - that would have to be a large error in our chunking logic if chunking is to blame.

**Comment by aakankshaduggal** [comment]
Thanks for chiming in @bbrowning. Have we tested the chunking for the knowledge documents associated with the knowledge yamls, I suspect those are relatively lengthier documents?

**Comment by bbrowning** [comment]
With #969 merged, this may fix this issue for any knowledge documents with long lines, as long as there are regular words in those lines and it's not just some giant random text without spaces.

For the original issue reported, it didn't sound like @HunterGerlach was teaching knowledge as it was only mentioned that the questions/answers were long, not the document. Were you able to trigger this with teaching skills using only long Q/A pairs, @HunterGerlach?

**Comment by HunterGerlach** [comment]
I was indeed teaching it knowledge. I can share the repo if needed (it's currently just the markdown from the InstructLab org + qna.yaml).

**Comment by gshipley** [comment]
it happened to me with a skill:

NFO 2024-04-30 20:45:01,424 generate_data.py:468 Selected taxonomy path compositional_skills->extraction->inference->quantitative->asciidoc->tables                                                                       
Generating dataset failed with the following error: There was a problem connecting to the server Error code: 400 - {'error': {'message': "This model's maximum context length is 4096 tokens. However, you requested 7332 t
okens (4260 in the messages, 3072 in the completion). Please reduce the length of the messages or completion.", 'type': 'invalid_request_error', 'param': 'messages', 'code': 'context_length_exceeded'}}             


Skill: https://github.com/gshipley/qna/blob/main/taxonomy/compositional_skills/extraction/inference/quantitative/asciidoc/tables/qna.yaml

**Comment by bbrowning** [comment]
@gshipley I haven't been able to reproduce the failure with that asciidoc qna.yaml yet, although will let it keep churning. In your error message, ilab managed to generate a prompt with 4260 tokens. Short-circuiting the generate logic and just dumping the number of tokens from each generated prompt in a loop, I've yet to construct a prompt with over 600 tokens. Something else might be going on here that would bloat this to 4260 tokens, but I haven't tracked it down yet.

**Comment by bbrowning** [comment]
Oh, I learned something new! We're feeding the synthetic generated instructions back into the model, so when we generate a long instruction and feed that back into the model we blow our context window. I had no idea we were treating earlier generated instructions as examples as well during the generation process. That feels perhaps a bit odd.

**Comment by bbrowning** [comment]
There area  few things to solve to fix this.

## Calculate tokens in each Q/A example before starting generation
Before starting generation, we should calculate (either exactly with a tokenizer, or approximately using a formula) the tokens of all their question and answer pairs and ensure it's under some threshold. If it's over, we should write out a sensible error message so the user knows what to fix. We should calculate that threshold dynamically based on the configured max context size of the model being served (can we ask that of the model server?), so that when a user raises the context size generate will automatically allow longer Q/A pairs.

## Response Length
https://github.com/instructlab/instructlab/blob/089c3617e6538ab070fed78ed44e337db3dfd60c/src/instructlab/generator/generate_data.py#L294-L296
We allow up to 3072 tokens for the completion response, so we can end up with a generated instruction almost this long. Both our prompt and the completion have to fit within the model context window, so we probably need to lower this value to something smaller or otherwise further restrict the allowed size of question/answer pair.

## Seed Instructions - human-written vs generated
https://github.com/instructlab/instructlab/blob/089c3617e6538ab070fed78ed44e337db3dfd60c/src/instructlab/generator/generate_data.py#L470-L474
We're feeding generating instruction back into the model as seed instructions. From reading the code history, this appears to perhaps be an unintended side-effect of some previous refactoring. I suggest we don't do that, as generated instructions can be quite long (see above) but also could progressively steer us further and further away from the intended topic and format as minor errors in generated data get reinforced over time during the generation run as more and more of our human-written seed examples get replaced with machine-generated ones.

**Comment by bbrowning** [comment]
I spoke with @abhi1092 elsewhere, and summarizing that learning here.

Adding synthetic instructions into our pool of examples is important for prompting diversity in the responses (see https://github.com/tatsu-lab/stanford_alpaca and https://github.com/yizhongw/self-instruct). So, we don't need to stop doing this, but both agreed that we should only insert generated responses below some token threshold as instructions, to eliminate the problem of long outliers blowing up our context window while retaining the benefits of prompting diversity.

**Comment by bbrowning** [comment]
#1069 should fix this for most cases, as it both validates the human-created examples are within a reasonable length compared to the server's context size and only adds generated responses into the pool of seed instructions if those generated responses are of a reasonable length compared to the server's context size.

**Comment by aakankshaduggal** [comment]
Thanks @bbrowning , really appreciate the quick response on this. Once this is merged, it should resolve the issues @gshipley and @HunterGerlach have been facing.

**Comment by AaronJaegerVA** [comment]
I just hit this today.

```
(venv-no-fips) (app-root) ilab system info
Platform:
  sys.version: 3.11.7 (main, Oct  9 2024, 00:00:00) [GCC 11.4.1 20231218 (Red Hat 11.4.1-3)]
  sys.platform: linux
  os.name: posix
  platform.release: 5.14.0-427.57.1.el9_4.x86_64
  platform.machine: x86_64
  platform.node: llm-test-0
  platform.python_version: 3.11.7
  os-release.ID: rhel
  os-release.VERSION_ID: 9.4
  os-release.PRETTY_NAME: Red Hat Enterprise Linux 9.4 (Plow)
  memory.total: 60.46 GB
  memory.available: 43.07 GB
  memory.used: 16.65 GB

InstructLab:
  instructlab.version: 0.24.3
  instructlab-dolomite.version: 0.2.0
  instructlab-eval.version: 0.5.1
  instructlab-quantize.version: 0.1.0
  instructlab-schema.version: 0.4.2
  instructlab-sdg.version: 0.7.3
  instructlab-training.version: 0.7.0

Torch:
  torch.version: 2.6.0+cu124
  torch.backends.cpu.capability: AVX2
  torch.version.cuda: 12.4
  torch.version.hip: None
  torch.cuda.available: True
  torch.backends.cuda.is_built: True
  torch.backends.mps.is_built: False
  torch.backends.mps.is_available: False
  torch.cuda.bf16: True
  torch.cuda.current.device: 0
  torch.cuda.0.name: NVIDIA L4
  torch.cuda.0.free: 21.8 GB
  torch.cuda.0.total: 22.0 GB
  torch.cuda.0.capability: 8.9 (see https://developer.nvidia.com/cuda-gpus#compute)

llama_cpp_python:
  llama_cpp_python.version: 0.3.6
  llama_cpp_python.supports_gpu_offload: False
```
